metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java

```diff
@@ -0,0 +1,232 @@
+import org.apache.kafka.common.metadata.AccessControlEntryRecord;
+import org.apache.kafka.common.metadata.RemoveAccessControlEntryRecord;
+import org.apache.kafka.common.requests.ApiError;
+import org.apache.kafka.metadata.authorizer.ClusterMetadataAuthorizer;
+import org.apache.kafka.metadata.authorizer.StandardAcl;
+import org.apache.kafka.metadata.authorizer.StandardAclRecordIterator;
+import org.apache.kafka.metadata.authorizer.StandardAclWithId;
+import org.apache.kafka.raft.OffsetAndEpoch;
+import org.apache.kafka.server.authorizer.AclCreateResult;
+import org.apache.kafka.server.authorizer.AclDeleteResult;
+import org.apache.kafka.server.authorizer.AclDeleteResult.AclBindingDeleteResult;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+import org.apache.kafka.timeline.SnapshotRegistry;
+import org.apache.kafka.timeline.TimelineHashMap;
+import org.apache.kafka.timeline.TimelineHashSet;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+
+
+/**
+ * The AclControlManager manages any ACLs that are stored in the __cluster_metadata topic.
+ * If the ACLs are stored externally (such as in ZooKeeper) then there will be nothing for
+ * this manager to do, and the authorizer field will always be Optional.empty.
+ *
+ * Because the Authorizer is being concurrently used by other threads, we need to be
+ * careful about snapshots. We don't want the Authorizer to act based on partial state
+ * during the loading process. Therefore, unlike most of the other managers,
+ * AclControlManager needs to receive callbacks when we start loading a snapshot and when
+ * we finish. The prepareForSnapshotLoad callback clears the authorizer field, preventing
+ * any changes from affecting the authorizer until completeSnapshotLoad is called.
+ * Note that the Authorizer's start() method will block until the first snapshot load has
+ * completed, which is another reason the prepare / complete callbacks are needed.
+ */
+public class AclControlManager {
+    private final TimelineHashMap<Uuid, StandardAcl> idToAcl;
+    private final TimelineHashSet<StandardAcl> existingAcls;
+    private final Optional<ClusterMetadataAuthorizer> authorizer;
+
+    AclControlManager(SnapshotRegistry snapshotRegistry,
+                      Optional<ClusterMetadataAuthorizer> authorizer) {
+        this.idToAcl = new TimelineHashMap<>(snapshotRegistry, 0);
+        this.existingAcls = new TimelineHashSet<>(snapshotRegistry, 0);
+        this.authorizer = authorizer;
+    }
+
+    ControllerResult<List<AclCreateResult>> createAcls(List<AclBinding> acls) {
+        List<AclCreateResult> results = new ArrayList<>(acls.size());
+        List<ApiMessageAndVersion> records = new ArrayList<>(acls.size());
+        for (AclBinding acl : acls) {
+            try {
+                validateNewAcl(acl);
```

Comment:
These validations seem to make sense. Currently they are all implemented by `AclAuthorizer`, so I wanted to ask whether it makes sense to pull them up. Basically are there any cases where a custom authorizer might be depending on any of the cases this validation is excluding?

Reply:
I think it would make sense to pull them up eventually in a follow-up PR.